File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.js
class Solution {
    candidates = []
    target = -1
    res = []
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) {
        this.candidates = candidates
        this.target = target

        this.candidates.sort((a, b) => a - b)

        this.foo(0, 0, [])
        return this.res
    }

    foo(idx, sum, state) {
        if(sum === this.target) {
            this.res.push([...state])
            return
        }

        for(let i = idx; i < this.candidates.length; i++) {
            if(i > idx && this.candidates[i] === this.candidates[i-1]) continue
            if(sum <= this.target) {
                state.push(this.candidates[i])
                this.foo(i + 1, sum + this.candidates[i], state)
                state.pop()
            }
        }
    } 
}

File: Data Structures & Algorithms/subsets-ii/submission-1.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> nums;

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        this->nums = nums;
        sort(this->nums.begin(), this->nums.end());
        vector<int> state;

        foo(0, state);
        return res;
    }

    void foo(int idx, vector<int>& state) {
        if(idx == nums.size()) {
            res.push_back(state);
            return;
        }

        state.push_back(nums[idx]);
        foo(idx+1, state);
        state.pop_back();
        while(idx+1 < nums.size() && nums[idx] == nums[idx+1]) {
            idx++;
        }
        foo(idx+1, state);
    }
};
